Source/WTF/wtf/Vector.h

```diff
@@ -108,217 +94,113 @@ struct VectorInitializer<true, false, T>
-    }
-
-    static void move(T* src, T* srcEnd, T* dst)
-    {
-        VectorMover<VectorTraits<T>::canMoveWithMemcpy, T>::move(src, srcEnd, dst);
-    }
 
     static void move(std::span<T> src, std::span<T> dst)
     {
-        VectorMover<VectorTraits<T>::canMoveWithMemcpy, T>::move(std::to_address(src.begin()), std::to_address(src.end()), std::to_address(dst.begin()));
+        move(std::to_address(src.begin()), std::to_address(src.end()), std::to_address(dst.begin()));
     }
 
     static void moveOverlapping(T* src, T* srcEnd, T* dst)
     {
-        VectorMover<VectorTraits<T>::canMoveWithMemcpy, T>::moveOverlapping(src, srcEnd, dst);
+        if constexpr (VectorTraits<T>::canMoveWithMemcpy)
+            memmove(static_cast<void*>(dst), static_cast<void*>(const_cast<T*>(src)), reinterpret_cast<const char*>(srcEnd) - reinterpret_cast<const char*>(src));
+        else {
+            if (src > dst)
+                move(src, srcEnd, dst);
+            else {
+                T* dstEnd = dst + (srcEnd - src);
+                while (src != srcEnd) {
+                    --srcEnd;
+                    --dstEnd;
+                    new (NotNull, dstEnd) T(WTFMove(*srcEnd));
+                    srcEnd->~T();
+                }
+            }
+        }
     }
 
     static void moveOverlapping(std::span<T> src, std::span<T> dst)
     {
-        VectorMover<VectorTraits<T>::canMoveWithMemcpy, T>::moveOverlapping(std::to_address(src.begin()), std::to_address(src.end()), std::to_address(dst.begin()));
+        moveOverlapping(std::to_address(src.begin()), std::to_address(src.end()), std::to_address(dst.begin()));
     }
 
     static void uninitializedCopy(const T* src, const T* srcEnd, T* dst)
     {
-        VectorCopier<VectorTraits<T>::canCopyWithMemcpy, T>::uninitializedCopy(src, srcEnd, dst);
+        if constexpr (VectorTraits<T>::canCopyWithMemcpy)
+            memcpy(static_cast<void*>(dst), static_cast<void*>(const_cast<T*>(src)), reinterpret_cast<const char*>(srcEnd) - reinterpret_cast<const char*>(src));
+        else {
+            while (src != srcEnd) {
+                new (NotNull, dst) T(*src);
+                ++dst;
+                ++src;
+            }
+        }
     }
 
     static void uninitializedCopy(std::span<const T> src, std::span<T> dst)
     {
-        VectorCopier<VectorTraits<T>::canCopyWithMemcpy, T>::uninitializedCopy(std::to_address(src.begin()), std::to_address(src.end()), std::to_address(dst.begin()));
+        uninitializedCopy(std::to_address(src.begin()), std::to_address(src.end()), std::to_address(dst.begin()));
+    }
+
+    template<typename U>
+    static void uninitializedImport(const U* src, const U* srcEnd, T* dst)
+    {
+        static_assert(!std::is_const_v<T>);
+        if constexpr (std::is_trivially_copyable_v<T> && std::is_same_v<T, U>)
```

Comment:
Seems like this won't ever be true because `U` is const and `T` is not. Are you sure the optimization is ever happening? How did you check that?

The old code checked `is_trivial` and this new code checks `is_trivially_copyable. Is this new check better? How did we verify it` doesn't remove an optimization or optimize incorrectly?

Reply:
> `U` is non-const.

I see. I guess I read the text wrong.

> `std::is_trivial` is deprecated in C++26.

That's interesting and it's forward looking to not sue something that will be deprecated in the future.

Does `is_trivially_copyable` give the exact sakes behavior or are there cases it will handle differently?